subsys/net/ip/net_if.c

```diff
@@ -4932,14 +4932,58 @@ void net_if_call_link_cb(struct net_if *iface, struct net_linkaddr *lladdr,
 	k_mutex_unlock(&lock);
 }
 
-static bool need_calc_checksum(struct net_if *iface, enum ethernet_hw_caps caps)
+static bool need_calc_checksum(struct net_if *iface, enum ethernet_hw_caps caps, enum net_if_checksum_protocol_type chksum_type)
 {
 #if defined(CONFIG_NET_L2_ETHERNET)
+	bool have_checksum_support;
+	bool have_checksum_support_protocol;
+	struct ethernet_config config;
+	int err;
+
+	have_checksum_support_protocol = false;
+
 	if (net_if_l2(iface) != &NET_L2_GET_NAME(ETHERNET)) {
 		return true;
 	}
 
-	return !(net_eth_get_hw_capabilities(iface) & caps);
+	have_checksum_support = net_eth_get_hw_capabilities(iface) & caps;
+	err = net_eth_get_hw_config(iface,caps == ETHERNET_HW_RX_CHKSUM_OFFLOAD ? ETHERNET_CONFIG_TYPE_RX_CHECKSUM_SUPPORT : ETHERNET_CONFIG_TYPE_TX_CHECKSUM_SUPPORT, &config);
+	if(err){
+		NET_DBG("Could not get ethernet config: %u",-err);
+		return true;
+	}
+
+	switch(chksum_type){
+		case NET_IF_CHECKSUM_PROTOCOL_IPV4_HEADER:
```

Comment:
Actually, not sure if we need a separate enum for net_if and ethernet, it just brings extra complexity to the solution.
I'd consider defining `enum net_if_checksum_type` as a bitmap at the net_if level, and just use it in ethernet layer. 
That way we could simply `return !(config.chksum_support & chksum_type);`

Reply:
See my comment above - it needs to be checked that both the IP and the transport protocol are supported.

Reply:
Ok, but we still could have a single enum (bitmap) for net_if and ethernet so we wouldn't need to translate here?

Reply:
I have implemented a version of this idea.
Because ethernet.h includes net_if.h, I needed to add a few extra defines to make this work, however.